File: problems/leetcode/0900/973.cpp
// 973. K Closest Points to Origin
// https://leetcode.com/problems/k-closest-points-to-origin/

using namespace std;

#include <queue>
#include <utility>
#include <vector>

typedef pair<long, int> pli; // (square of distance, index)
typedef long long ll;

class Solution {
public:
  vector<vector<int>> kClosest(vector<vector<int>> &points, int k) {
    vector<vector<int>> answer;
    int length = points.size();

    priority_queue<pli, vector<pli>, greater<pli>> pq;

    for (int i = 0; i < length; i++) {
      pq.push({squareOf(points[i][0]) + squareOf(points[i][1]), i});
    }

    for (int i = 0; i < k; i++) {
      answer.push_back(points[pq.top().second]);
      pq.pop();
    }

    return answer;
  }

  ll squareOf(int n) { return n * n; }
};

File: problems/leetcode/0900/973/973.py
# 973. K Closest Points to Origin
# https://leetcode.com/problems/k-closest-points-to-origin/description/

from collections import heapq
from typing import *

class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        closestPoints = [] # heapq - (square distance, x, y)

        for x, y in points:
            heapq.heappush(closestPoints, (-(x ** 2 + y ** 2), x, y))
            if len(closestPoints) > k:
                heapq.heappop(closestPoints)

        return [[x, y] for _, x, y in closestPoints]
